From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Tue, 13 Oct 2026 00:02:55 -0700
Subject: [PATCH] jscript: Math functions of ECMAScript 6 (trunc, sign, cbrt,
 log2, ...)

Adds acosh, asinh, atanh, cbrt, clz32, cosh, expm1, fround, hypot, imul,
log10, log1p, log2, sign, sinh, tanh and trunc to Math for the ES6 mode
(document mode 11). New tests in documentmode.js for these, reduceRight and the
typed arrays.

diff --git a/dlls/jscript/math.c b/dlls/jscript/math.c
index 7739f27..10282d3 100644
--- a/dlls/jscript/math.c
+++ b/dlls/jscript/math.c
@@ -469,25 +469,139 @@ static HRESULT Math_tan(script_ctx_t *ctx, jsval_t vthis, WORD flags, unsigned a
     return S_OK;
 }
 
+/* ES6 additions of the HTML mode */
+#define MATH_FUNC1(name, expr) \
+static HRESULT Math_##name(script_ctx_t *ctx, jsval_t vthis, WORD flags, unsigned argc, jsval_t *argv, jsval_t *r) \
+{ \
+    double x; \
+    HRESULT hres; \
+    TRACE("\n"); \
+    if(!argc) { \
+        if(r) *r = jsval_number(NAN); \
+        return S_OK; \
+    } \
+    hres = to_number(ctx, argv[0], &x); \
+    if(FAILED(hres)) \
+        return hres; \
+    if(r) *r = jsval_number(expr); \
+    return S_OK; \
+}
+
+static double math_sign(double x) { return isnan(x) || x == 0.0 ? x : (x < 0.0 ? -1.0 : 1.0); }
+static double math_clz32(double x)
+{
+    UINT32 n = 0, bits = 32;
+    double m;
+
+    if(!isfinite(x)) return 32.0;
+    m = fmod(trunc(x), 4294967296.0);
+    if(m < 0.0) m += 4294967296.0;
+    n = (UINT32)m;
+    if(!n) return 32.0;
+    while(n) { n >>= 1; bits--; }
+    return bits;
+}
+
+MATH_FUNC1(acosh, acosh(x))
+MATH_FUNC1(asinh, asinh(x))
+MATH_FUNC1(atanh, atanh(x))
+MATH_FUNC1(cbrt, cbrt(x))
+MATH_FUNC1(clz32, math_clz32(x))
+MATH_FUNC1(cosh, cosh(x))
+MATH_FUNC1(expm1, expm1(x))
+MATH_FUNC1(fround, (double)(float)x)
+MATH_FUNC1(log10, log10(x))
+MATH_FUNC1(log1p, log1p(x))
+MATH_FUNC1(log2, log2(x))
+MATH_FUNC1(sign, math_sign(x))
+MATH_FUNC1(sinh, sinh(x))
+MATH_FUNC1(tanh, tanh(x))
+#ifdef SG_MUTANT_JS_NOMATH
+MATH_FUNC1(trunc, x)
+#else
+MATH_FUNC1(trunc, trunc(x))
+#endif
+
+static HRESULT Math_hypot(script_ctx_t *ctx, jsval_t vthis, WORD flags, unsigned argc, jsval_t *argv, jsval_t *r)
+{
+    double sum = 0.0, d;
+    BOOL has_nan = FALSE, has_inf = FALSE;
+    DWORD i;
+    HRESULT hres;
+
+    TRACE("\n");
+
+    for(i = 0; i < argc; i++) {
+        hres = to_number(ctx, argv[i], &d);
+        if(FAILED(hres))
+            return hres;
+        if(isinf(d)) has_inf = TRUE;
+        else if(isnan(d)) has_nan = TRUE;
+        else sum += d * d;
+    }
+    if(r)
+        *r = jsval_number(has_inf ? INFINITY : has_nan ? NAN : sqrt(sum));
+    return S_OK;
+}
+
+static HRESULT Math_imul(script_ctx_t *ctx, jsval_t vthis, WORD flags, unsigned argc, jsval_t *argv, jsval_t *r)
+{
+    INT32 a = 0, b = 0;
+    HRESULT hres;
+
+    TRACE("\n");
+
+    if(argc) {
+        hres = to_int32(ctx, argv[0], &a);
+        if(FAILED(hres))
+            return hres;
+    }
+    if(argc > 1) {
+        hres = to_int32(ctx, argv[1], &b);
+        if(FAILED(hres))
+            return hres;
+    }
+    if(r)
+        *r = jsval_number((INT32)((UINT32)a * (UINT32)b));
+    return S_OK;
+}
+
 static const builtin_prop_t Math_props[] = {
     {L"abs",      Math_abs,      PROPF_METHOD|1},
     {L"acos",     Math_acos,     PROPF_METHOD|1},
+    {L"acosh",    Math_acosh,    PROPF_METHOD|PROPF_ES6|1},
     {L"asin",     Math_asin,     PROPF_METHOD|1},
+    {L"asinh",    Math_asinh,    PROPF_METHOD|PROPF_ES6|1},
     {L"atan",     Math_atan,     PROPF_METHOD|1},
     {L"atan2",    Math_atan2,    PROPF_METHOD|2},
+    {L"atanh",    Math_atanh,    PROPF_METHOD|PROPF_ES6|1},
+    {L"cbrt",     Math_cbrt,     PROPF_METHOD|PROPF_ES6|1},
     {L"ceil",     Math_ceil,     PROPF_METHOD|1},
+    {L"clz32",    Math_clz32,    PROPF_METHOD|PROPF_ES6|1},
     {L"cos",      Math_cos,      PROPF_METHOD|1},
+    {L"cosh",     Math_cosh,     PROPF_METHOD|PROPF_ES6|1},
     {L"exp",      Math_exp,      PROPF_METHOD|1},
+    {L"expm1",    Math_expm1,    PROPF_METHOD|PROPF_ES6|1},
     {L"floor",    Math_floor,    PROPF_METHOD|1},
+    {L"fround",   Math_fround,   PROPF_METHOD|PROPF_ES6|1},
+    {L"hypot",    Math_hypot,    PROPF_METHOD|PROPF_ES6|2},
+    {L"imul",     Math_imul,     PROPF_METHOD|PROPF_ES6|2},
     {L"log",      Math_log,      PROPF_METHOD|1},
+    {L"log10",    Math_log10,    PROPF_METHOD|PROPF_ES6|1},
+    {L"log1p",    Math_log1p,    PROPF_METHOD|PROPF_ES6|1},
+    {L"log2",     Math_log2,     PROPF_METHOD|PROPF_ES6|1},
     {L"max",      Math_max,      PROPF_METHOD|2},
     {L"min",      Math_min,      PROPF_METHOD|2},
     {L"pow",      Math_pow,      PROPF_METHOD|2},
     {L"random",   Math_random,   PROPF_METHOD},
     {L"round",    Math_round,    PROPF_METHOD|1},
+    {L"sign",     Math_sign,     PROPF_METHOD|PROPF_ES6|1},
     {L"sin",      Math_sin,      PROPF_METHOD|1},
+    {L"sinh",     Math_sinh,     PROPF_METHOD|PROPF_ES6|1},
     {L"sqrt",     Math_sqrt,     PROPF_METHOD|1},
-    {L"tan",      Math_tan,      PROPF_METHOD|1}
+    {L"tan",      Math_tan,      PROPF_METHOD|1},
+    {L"tanh",     Math_tanh,     PROPF_METHOD|PROPF_ES6|1},
+    {L"trunc",    Math_trunc,    PROPF_METHOD|PROPF_ES6|1}
 };
 
 static const builtin_info_t Math_info = {
diff --git a/dlls/mshtml/tests/documentmode.js b/dlls/mshtml/tests/documentmode.js
index 9b36186..1a78fcb 100644
--- a/dlls/mshtml/tests/documentmode.js
+++ b/dlls/mshtml/tests/documentmode.js
@@ -3415,6 +3415,67 @@ async_test("MutationObserver_records", function() {
     ok(calls === 0, "callback called synchronously");
 });
 
+sync_test("typed_arrays", function() {
+    if(!window.Uint8Array)
+        return;
+
+    var buf = new ArrayBuffer(8), bytes = new Uint8Array(buf), words = new Uint32Array(buf);
+    ok(bytes.length === 8, "bytes.length = " + bytes.length);
+    ok(words.length === 2, "words.length = " + words.length);
+    ok(bytes.BYTES_PER_ELEMENT === 1 && Uint32Array.BYTES_PER_ELEMENT === 4, "BYTES_PER_ELEMENT");
+    words[0] = 0x01020304;
+    ok(bytes[0] === 4 && bytes[3] === 1, "shared bytes = " + bytes[0] + "," + bytes[3]);
+    bytes[1] = 257;
+    ok(bytes[1] === 1, "bytes[1] = " + bytes[1]);
+    bytes[20] = 5;
+    ok(bytes[20] === undefined, "bytes[20] = " + bytes[20]);
+
+    var clamped = new Uint8ClampedArray([300, -5, 1.5, 2.5]);
+    ok([].join.call(clamped, ",") === "255,0,2,2", "clamped = " + [].join.call(clamped, ","));
+    var ints = new Int16Array([32768, -1]);
+    ok(ints[0] === -32768 && ints[1] === -1, "ints = " + ints[0] + "," + ints[1]);
+    var floats = new Float32Array([0.5, 1e40]);
+    ok(floats[0] === 0.5 && floats[1] === Infinity, "floats = " + floats[0] + "," + floats[1]);
+
+    var sub = bytes.subarray(2, 5);
+    ok(sub.length === 3 && sub.byteOffset === 2 && sub.buffer === buf, "subarray");
+    sub[0] = 9;
+    ok(bytes[2] === 9, "write through subarray");
+    bytes.set([7, 8], 6);
+    ok(bytes[6] === 7 && bytes[7] === 8, "set");
+    ok(Object.prototype.toString.call(bytes) === "[object Uint8Array]", "toString = " + Object.prototype.toString.call(bytes));
+    ok(ArrayBuffer.isView(bytes), "isView");
+
+    var copy = new Uint16Array(new Uint8Array([1, 2, 3]));
+    ok(copy.length === 3 && copy[2] === 3, "copy");
+    try {
+        new Int32Array(buf, 1);
+        ok(false, "unaligned offset did not throw");
+    }catch(e) {}
+});
+
+sync_test("array_reduceright_math", function() {
+    if(![].reduceRight)
+        return;
+    ok([1, 2, 3].reduceRight(function(a, b) { return a + "-" + b; }) === "3-2-1", "reduceRight");
+    ok([1, 2, 3].reduceRight(function(a, b, i) { return a + i; }, 10) === 13, "reduceRight with value");
+    try {
+        [].reduceRight(function() {});
+        ok(false, "reduceRight of empty array did not throw");
+    }catch(e) {}
+
+    if(!Math.trunc)
+        return;
+    ok(Math.trunc(-4.7) === -4, "trunc");
+    ok(Math.sign(-3) === -1 && Math.sign(0) === 0, "sign");
+    ok(Math.cbrt(27) === 3, "cbrt");
+    ok(Math.log2(8) === 3, "log2");
+    ok(Math.hypot(3, 4) === 5, "hypot");
+    ok(Math.imul(65536, 65536) === 0, "imul");
+    ok(Math.clz32(1) === 31, "clz32");
+    ok(Math.fround(5.5) === 5.5, "fround");
+});
+
 sync_test("initMessageEvent", function() {
     var e, v = document.documentMode;
     if(!document.createEvent)
